g419-liner2-core/src/main/java/g419/liner2/core/features/tokens/TfIdfFeature.java

```diff
@@ -1,38 +1,38 @@
 package g419.liner2.core.features.tokens;
 
+import g419.liner2.core.tools.BaseCount;
+import g419.corpus.structure.Document;
+import g419.corpus.structure.Paragraph;
 import g419.corpus.structure.Sentence;
 import g419.corpus.structure.Token;
 
 import java.util.List;
 
 
-public class TfIdfFeature extends TokenInSentenceFeature {
+public class TfIdfFeature extends TokenInDocumentFeature {
 
   public TfIdfFeature(String name) {
-    super(name);
+      super(name);
   }
 
-
   @Override
-  public void generate(Sentence sentence) {
-
-    int thisFeatureIdx = sentence.getAttributeIndex().getIndex(this.getName());
-    List<Token> tokens = sentence.getTokens();
-
-    float tokenNumber = (float) (sentence.getDocument().getTokenNumber());
-
-    int tokenIdx = 0;
-    while (tokenIdx < sentence.getTokenNumber()) {
-      Token t = tokens.get(tokenIdx);
-      String base = t.getAttributeValue("base");
-      float idf = Float.parseFloat(t.getAttributeValue("key_value_idf_base"));
-      float tf = (float) (sentence.getDocument().getBaseCount(base));
-      float normTf = tf / tokenNumber;
-      float tfIdf = normTf * idf;
-      t.setAttributeValue(thisFeatureIdx, tfIdf > 0.05 ? "1" : "0");
-      //String.format("%.2f", tfIdf)
-      tokenIdx++;
+  public void generate(Document document) {
+    BaseCount baseCount = new BaseCount(document);
+    int thisFeatureIdx = document.getAttributeIndex().getIndex(this.getName());
+
+    float tokenNumber = (float) document.getTokenNumber();
+    
+    for(Paragraph paragraph : document.getParagraphs()){
+      for(Sentence sentence : paragraph.getSentences()){
+        for(Token token : sentence.getTokens()){
+	      final String base = token.getAttributeValue("base");
+	      float idf = Float.parseFloat(token.getAttributeValue("key_value_idf_base"));
```

Comment:
W przypadku braku cechy `key_value_idf_base` zostanie tutaj rzucony wyjątek `OutOfBoundaryException`, który nie ułatwi debufowania. Nie lepiej ponownie wyliczyć idf, aby nie uzależniać tej cechy od generatora dla `key_value_idf_base`?

Reply:
Można, tylko nie wiem czym jest idf. Domyślam się, że tf to token frequency, a czym jest idf? https://en.wikipedia.org/wiki/Tf%E2%80%93idf#Inverse_document_frequency
Jeśli to to czyli "It is the logarithmically scaled inverse fraction of the documents that contain the word (obtained by dividing the total number of documents by the number of documents containing the term, and then taking the logarithm of that quotient) [...]", to czy liczymy to na dokumentach (nie widzę opcji dostępu do wielu dokumentów, więc chyba nie) czy na paragrafach?

Reply:
To są pozostałości po pewnym eksperymencie, które można usunąć. Wartości idf były czytane z pliku. Do wyrzucenia: plik TfIdfFeature.java oraz obsługa cechy w g419-liner2-core/src/main/java/g419/liner2/core/features/tokens/TokenFeatureFactory.java